pkg/apis/pipeline/v1beta1/pipeline_validation.go

```diff
@@ -91,6 +91,61 @@ func (ps *PipelineSpec) Validate(ctx context.Context) (errs *apis.FieldError) {
 	errs = errs.Also(validateMatrix(ctx, ps.Tasks).ViaField("tasks"))
 	errs = errs.Also(validateMatrix(ctx, ps.Finally).ViaField("finally"))
 	errs = errs.Also(validateResultsFromMatrixedPipelineTasksNotConsumed(ps.Tasks, ps.Finally))
+	errs = errs.Also(ps.ValidateBetaFields(ctx))
+	return errs
+}
+
+// ValidateBetaFields returns an error if the Pipeline spec uses beta features but does not
+// have "enable-api-fields" set to "alpha" or "beta".
```

Comment:
do you think it is ok to add a link to the list of our beta features here? So we can make sure the validation here covers all the features

Reply:
Thanks for the suggestion @Yongxuanzhang  
I think the point for this func is to cover all existing beta features related with the pipelineSpec that are coupled before (the beta features that had different validations in v1 versus v1beta1). So not essentially all beta features here.

I understand the point here is to give users/cluster operators full notice of the change ahead with the transition to per-feature flag.
I've put a list in https://github.com/tektoncd/community/blob/main/teps/0138-decouple-api-and-feature-versioning.md as a cutoff of all existing beta/alpha features that are going to stay behind `enable-api-fields` and such validations. Does that sound good to add them in the api-compatibility policy update PR with Per-feature flag introduced to give heads up since this PR mainly handles the decoupling?

Reply:
Oh I think I misunderstood the code here, I thought it will cover all but the pipelineref is in another function 😂 I think it's fine.